In [1]:
%load_ext autoreload 
%autoreload 2

In [2]:
import torch 
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import networkx as nx
from collections import Counter, defaultdict
from dataclasses import dataclass
from pathlib import Path
from typing import Dict, List, MutableMapping, Tuple
import os
os.environ["CUDA_VISIBLE_DEVICES"] = "2"

In [4]:
import sys
import os
PROJECT_ROOT = os.path.abspath(os.path.join(os.getcwd(), ".."))
if PROJECT_ROOT not in sys.path:
    sys.path.insert(0, PROJECT_ROOT)
from src.align import *
from src.evaluate import *
from src.features import *
from src.process import *
from src.utils import *
from src.vis import *

In [21]:
df = pd.read_csv("outputs/eval_all.csv")
df_CG = df[df['pair_id'].str.contains('CG')]
df_DMR = df[df['pair_id'].str.contains('DMR')]
df_DMC = df[df['pair_id'].str.contains('DMC')]
df_STICKY = df[df['pair_id'].str.contains('STICKY')]

# print("\ndf_CG mean:\n", df_CG.mean(numeric_only=True))
# print("\ndf_DMR mean:\n", df_DMR.mean(numeric_only=True))
# print("\ndf_DMC mean:\n", df_DMC.mean(numeric_only=True))
# print("\ndf_STICKY mean:\n", df_STICKY.mean(numeric_only=True))
means_df = pd.DataFrame({
    'CG': df_CG.mean(numeric_only=True),
    'DMR': df_DMR.mean(numeric_only=True),
    'DMC': df_DMC.mean(numeric_only=True),
    'STICKY': df_STICKY.mean(numeric_only=True)
})
print(means_df)

                                    CG         DMR         DMC      STICKY
sim_soft                    186.268817  187.788711  148.639532  123.373556
sim_hard_source             141.306784  166.800421  143.172753  108.164595
sim_hard_target             132.893560  162.119218  140.239204  104.286737
fo_consistency_soft           0.909478    0.913486    0.721979    0.598252
fo_consistency_hard_source    0.944327    0.839665    0.720465    0.543438
fo_consistency_hard_target    0.940190    0.820040    0.709806    0.525713
topk_recall@3                 0.950026    0.846821    0.770625    0.579690
topk_recall@5                 0.950170    0.847358    0.772347    0.583133
topk_recall@10                0.950448    0.848272    0.773760    0.585945
s3_hard                       0.632701    0.686250    0.607822    0.541511
total_loss                    0.136777    0.105469    0.079098    0.111527
runtime_seconds             136.232199   25.358272   97.521741   57.124590


In [ ]:
df1 = pd.read_csv("IsoRankN/napa_eval_outputs/eval_all_1.csv")
df1_CG = df1[df1['pair_id'].str.contains('CG')]
df1_DMR = df1[df1['pair_id'].str.contains('DMR')]
df1_DMC = df1[df1['pair_id'].str.contains('DMC')]
df1_STICKY = df1[df1['pair_id'].str.contains('STICKY')]

means_df1 = pd.DataFrame({
    'CG': df1_CG.mean(numeric_only=True),
    'DMR': df1_DMR.mean(numeric_only=True),
    'DMC': df1_DMC.mean(numeric_only=True),
    'STICKY': df1_STICKY.mean(numeric_only=True)
})
print(means_df1)

                                    CG         DMR         DMC      STICKY
sim_soft                     53.703937   73.548540   74.847611   80.537824
sim_hard_source             184.988310  224.434677  221.556172  220.591169
sim_hard_target             183.085904  226.614720  223.428217  224.643245
fo_consistency_soft           0.069994    0.239653    0.249898    0.279539
fo_consistency_hard_source    0.838795    0.933908    0.934052    0.929245
fo_consistency_hard_target    0.833719    0.910113    0.914524    0.900684
topk_recall@3                 0.916012    0.979553    0.981901    0.979725
topk_recall@5                 0.942520    0.990038    0.990868    0.988638
topk_recall@10                0.977681    0.995787    0.996061    0.993717
s3_hard                       0.342723    0.594458    0.541914    0.480919
total_loss                         NaN         NaN         NaN         NaN
runtime_seconds              32.000000   48.100000   40.700000   37.800000


In [22]:
df2 = pd.read_csv("IsoRankN/napa_eval_outputs/eval_all_2.csv")
df2_CG = df2[df2['pair_id'].str.contains('CG')]
df2_DMR = df2[df2['pair_id'].str.contains('DMR')]
df2_DMC = df2[df2['pair_id'].str.contains('DMC')]
df2_STICKY = df2[df2['pair_id'].str.contains('STICKY')]

means_df2 = pd.DataFrame({
    'CG': df2_CG.mean(numeric_only=True),
    'DMR': df2_DMR.mean(numeric_only=True),
    'DMC': df2_DMC.mean(numeric_only=True),
    'STICKY': df2_STICKY.mean(numeric_only=True)
})
print(means_df2)

                                    CG         DMR         DMC      STICKY
sim_soft                    196.106821  221.261580  217.151715  218.304236
sim_hard_source             155.059496  179.487263  178.480795  185.752794
sim_hard_target             161.294280  197.362257  194.406778  197.087326
fo_consistency_soft           0.779899    0.922606    0.918755    0.924806
fo_consistency_hard_source    0.843820    0.775082    0.780989    0.810319
fo_consistency_hard_target    0.868478    0.843471    0.842071    0.852045
topk_recall@3                 0.864220    0.814139    0.816213    0.838565
topk_recall@5                 0.867023    0.814831    0.816833    0.839903
topk_recall@10                0.868234    0.816131    0.817894    0.843085
s3_hard                       0.376845    0.486553    0.411078    0.456432
total_loss                         NaN         NaN         NaN         NaN
runtime_seconds              32.000000   48.100000   40.700000   37.800000


In [23]:
df3 = pd.read_csv("IsoRankN/napa_eval_outputs/eval_all_3.csv")
df3_CG = df3[df3['pair_id'].str.contains('CG')]
df3_DMR = df3[df3['pair_id'].str.contains('DMR')]
df3_DMC = df3[df3['pair_id'].str.contains('DMC')]
df3_STICKY = df3[df3['pair_id'].str.contains('STICKY')]

means_df3 = pd.DataFrame({
    'CG': df3_CG.mean(numeric_only=True),
    'DMR': df3_DMR.mean(numeric_only=True),
    'DMC': df3_DMC.mean(numeric_only=True),
    'STICKY': df3_STICKY.mean(numeric_only=True)
})
print(means_df3)

                                   CG        DMR        DMC     STICKY
sim_soft                    10.139926   2.525271   5.109611   2.695185
sim_hard_source             34.579334  14.681424  15.720496  12.349217
sim_hard_target             28.913787  14.286720  14.832476  12.614122
fo_consistency_soft          0.004220   0.011232   0.023236   0.011743
fo_consistency_hard_source   0.054483   0.073997   0.076842   0.057500
fo_consistency_hard_target   0.063817   0.071033   0.073147   0.058785
topk_recall@3                0.132985   0.151474   0.199729   0.102073
topk_recall@5                0.197199   0.199897   0.290098   0.129971
topk_recall@10               0.330259   0.276213   0.432843   0.175939
s3_hard                      0.027957   0.193036   0.095310   0.092629
total_loss                        NaN        NaN        NaN        NaN
runtime_seconds             32.600000  48.000000  41.300000  39.300000


In [24]:
df4 = pd.read_csv("IsoRankN/napa_eval_outputs/eval_all_4.csv")
df4_CG = df4[df4['pair_id'].str.contains('CG')]
df4_DMR = df4[df4['pair_id'].str.contains('DMR')]
df4_DMC = df4[df4['pair_id'].str.contains('DMC')]
df4_STICKY = df4[df4['pair_id'].str.contains('STICKY')]

means_df4 = pd.DataFrame({
    'CG': df4_CG.mean(numeric_only=True),
    'DMR': df4_DMR.mean(numeric_only=True),
    'DMC': df4_DMC.mean(numeric_only=True),
    'STICKY': df4_STICKY.mean(numeric_only=True)
})
print(means_df4)

                                   CG        DMR        DMC     STICKY
sim_soft                    73.301590  81.979840  73.659525  79.785535
sim_hard_source             34.443584  25.012425  31.218463  13.667519
sim_hard_target             33.365789  27.160480  34.744639  15.077511
fo_consistency_soft          0.339636   0.395910   0.351703   0.369076
fo_consistency_hard_source   0.219708   0.125300   0.154637   0.064742
fo_consistency_hard_target   0.223418   0.135908   0.172165   0.072366
topk_recall@3                0.240151   0.141433   0.176714   0.081098
topk_recall@5                0.242844   0.144577   0.180726   0.086457
topk_recall@10               0.245507   0.150211   0.186455   0.098102
s3_hard                      0.097950   0.074452   0.071178   0.051901
total_loss                        NaN        NaN        NaN        NaN
runtime_seconds             32.600000  48.000000  41.300000  39.300000


#### test

In [51]:
category = "STICKY"
TIME = "20260104_140536"
import src.evaluate as eval
family = load_family_data(Path(f"data/NAPAbench/NAPAbenchVer2/benchmarkDataset/2way/{category}/family1"))
sim_mat = eval._similarity_matrix_from_file(f"data/NAPAbench/NAPAbenchVer2/benchmarkDataset/2way/{category}/family1/A-B.sim", 
source_nodes=list(family.graphs["A"].nodes()), target_nodes=list(family.graphs["B"].nodes()))

2026-01-10 16:39:04,118 | INFO | src.process | Loaded A (3000 nodes, 22109 edges)
2026-01-10 16:39:04,326 | INFO | src.process | Loaded B (4000 nodes, 30737 edges)
2026-01-10 16:39:04,365 | INFO | src.process | Loaded A annotations (2901 nodes with FO labels)
2026-01-10 16:39:04,369 | INFO | src.process | Loaded B annotations (3760 nodes with FO labels)
2026-01-10 16:39:04,634 | INFO | src.process | Loaded similarity file A-B (29942 pairs)


In [ ]:
feature_path = f"outputs/2way/{category}/family1/A_vs_B/{TIME}"
output_dir = "/home/bcl/wanghongyu/other/networkscience/outputs"
align_config = AlignmentConfig(
    alpha=0.9,
    rho=0.1,
    eps=1e-5,
    solver="sinkhorn",
    solver_params={"nits_bcd": 25, "nits_uot": 1000, "tol_bcd": 1e-4, "tol_uot": 1e-10},
    device="cuda",
    verbose=True
)
result = fugw_align(family.graphs["A"], family.graphs["B"], 
feature_path=feature_path, 
align_config=align_config, 
output_dir=output_dir,
sim_mat = sim_mat)

source_nodes = result['source_nodes']
target_nodes = result['target_nodes']
pi = result["pi"].numpy() / result["pi"].numpy().sum()

2026-01-10 16:53:52,543 | INFO | src.align | Loaded cached features from outputs/2way/STICKY/family1/A_vs_B/20260104_140536
2026-01-10 16:53:52,544 | INFO | src.align | Source features: (3000, 5), Target features: (4000, 5)
2026-01-10 16:53:52,544 | INFO | src.align | Source geometry: (3000, 3000), Target geometry: (4000, 4000)


Using cost matrix


[16:53:54] Validation data for feature maps is not provided. Using training data instead.              ]8;id=611880;file:///home/bcl/wanghongyu/.conda/envs/network/lib/python3.11/site-packages/fugw/mappings/dense.py\dense.py]8;;\:]8;id=372297;file:///home/bcl/wanghongyu/.conda/envs/network/lib/python3.11/site-packages/fugw/mappings/dense.py#204\204]8;;\

           Validation data for anatomical kernels is not provided. Using training data instead.        ]8;id=397590;file:///home/bcl/wanghongyu/.conda/envs/network/lib/python3.11/site-packages/fugw/mappings/dense.py\dense.py]8;;\:]8;id=385351;file:///home/bcl/wanghongyu/.conda/envs/network/lib/python3.11/site-packages/fugw/mappings/dense.py#231\231]8;;\

Output()

Output()

[16:54:37] BCD step 1/25   FUGW loss:      0.6531960368156433      Validation loss:                    ]8;id=22873;file:///home/bcl/wanghongyu/.conda/envs/network/lib/python3.11/site-packages/fugw/solvers/dense.py\dense.py]8;;\:]8;id=457083;file:///home/bcl/wanghongyu/.conda/envs/network/lib/python3.11/site-packages/fugw/solvers/dense.py#521\521]8;;\
           0.6531960368156433                                                                                      

Output()

Output()

[16:55:21] BCD step 2/25   FUGW loss:      0.6225553750991821      Validation loss:                    ]8;id=369094;file:///home/bcl/wanghongyu/.conda/envs/network/lib/python3.11/site-packages/fugw/solvers/dense.py\dense.py]8;;\:]8;id=200181;file:///home/bcl/wanghongyu/.conda/envs/network/lib/python3.11/site-packages/fugw/solvers/dense.py#521\521]8;;\
           0.6225553750991821                                                                                      

Output()

Output()

[16:56:04] BCD step 3/25   FUGW loss:      0.48329469561576843     Validation loss:                    ]8;id=183971;file:///home/bcl/wanghongyu/.conda/envs/network/lib/python3.11/site-packages/fugw/solvers/dense.py\dense.py]8;;\:]8;id=264473;file:///home/bcl/wanghongyu/.conda/envs/network/lib/python3.11/site-packages/fugw/solvers/dense.py#521\521]8;;\
           0.48329469561576843                                                                                     

Output()

Output()

[16:56:48] BCD step 4/25   FUGW loss:      0.40758955478668213     Validation loss:                    ]8;id=170409;file:///home/bcl/wanghongyu/.conda/envs/network/lib/python3.11/site-packages/fugw/solvers/dense.py\dense.py]8;;\:]8;id=739422;file:///home/bcl/wanghongyu/.conda/envs/network/lib/python3.11/site-packages/fugw/solvers/dense.py#521\521]8;;\
           0.40758955478668213                                                                                     

Output()

Output()

[16:57:32] BCD step 5/25   FUGW loss:      0.3691610097885132      Validation loss:                    ]8;id=567512;file:///home/bcl/wanghongyu/.conda/envs/network/lib/python3.11/site-packages/fugw/solvers/dense.py\dense.py]8;;\:]8;id=784920;file:///home/bcl/wanghongyu/.conda/envs/network/lib/python3.11/site-packages/fugw/solvers/dense.py#521\521]8;;\
           0.3691610097885132                                                                                      

Output()

Output()

[16:58:15] BCD step 6/25   FUGW loss:      0.3425094783306122      Validation loss:                    ]8;id=792053;file:///home/bcl/wanghongyu/.conda/envs/network/lib/python3.11/site-packages/fugw/solvers/dense.py\dense.py]8;;\:]8;id=469820;file:///home/bcl/wanghongyu/.conda/envs/network/lib/python3.11/site-packages/fugw/solvers/dense.py#521\521]8;;\
           0.3425094783306122                                                                                      

Output()

Output()

[16:58:59] BCD step 7/25   FUGW loss:      0.32096683979034424     Validation loss:                    ]8;id=376918;file:///home/bcl/wanghongyu/.conda/envs/network/lib/python3.11/site-packages/fugw/solvers/dense.py\dense.py]8;;\:]8;id=513798;file:///home/bcl/wanghongyu/.conda/envs/network/lib/python3.11/site-packages/fugw/solvers/dense.py#521\521]8;;\
           0.32096683979034424                                                                                     

Output()

Output()

[16:59:43] BCD step 8/25   FUGW loss:      0.30911701917648315     Validation loss:                    ]8;id=486330;file:///home/bcl/wanghongyu/.conda/envs/network/lib/python3.11/site-packages/fugw/solvers/dense.py\dense.py]8;;\:]8;id=85101;file:///home/bcl/wanghongyu/.conda/envs/network/lib/python3.11/site-packages/fugw/solvers/dense.py#521\521]8;;\
           0.30911701917648315                                                                                     

Output()

Output()

[17:00:27] BCD step 9/25   FUGW loss:      0.2990916669368744      Validation loss:                    ]8;id=47730;file:///home/bcl/wanghongyu/.conda/envs/network/lib/python3.11/site-packages/fugw/solvers/dense.py\dense.py]8;;\:]8;id=863703;file:///home/bcl/wanghongyu/.conda/envs/network/lib/python3.11/site-packages/fugw/solvers/dense.py#521\521]8;;\
           0.2990916669368744                                                                                      

Output()

Output()

[17:01:11] BCD step 10/25  FUGW loss:      0.29168134927749634     Validation loss:                    ]8;id=250903;file:///home/bcl/wanghongyu/.conda/envs/network/lib/python3.11/site-packages/fugw/solvers/dense.py\dense.py]8;;\:]8;id=551846;file:///home/bcl/wanghongyu/.conda/envs/network/lib/python3.11/site-packages/fugw/solvers/dense.py#521\521]8;;\
           0.29168134927749634                                                                                     

Output()

Output()

[17:01:55] BCD step 11/25  FUGW loss:      0.2862621545791626      Validation loss:                    ]8;id=414138;file:///home/bcl/wanghongyu/.conda/envs/network/lib/python3.11/site-packages/fugw/solvers/dense.py\dense.py]8;;\:]8;id=289930;file:///home/bcl/wanghongyu/.conda/envs/network/lib/python3.11/site-packages/fugw/solvers/dense.py#521\521]8;;\
           0.2862621545791626                                                                                      

Output()

Output()

[17:02:38] BCD step 12/25  FUGW loss:      0.28040674328804016     Validation loss:                    ]8;id=751061;file:///home/bcl/wanghongyu/.conda/envs/network/lib/python3.11/site-packages/fugw/solvers/dense.py\dense.py]8;;\:]8;id=506405;file:///home/bcl/wanghongyu/.conda/envs/network/lib/python3.11/site-packages/fugw/solvers/dense.py#521\521]8;;\
           0.28040674328804016                                                                                     

Output()

Output()

[17:03:22] BCD step 13/25  FUGW loss:      0.27642786502838135     Validation loss:                    ]8;id=310775;file:///home/bcl/wanghongyu/.conda/envs/network/lib/python3.11/site-packages/fugw/solvers/dense.py\dense.py]8;;\:]8;id=362001;file:///home/bcl/wanghongyu/.conda/envs/network/lib/python3.11/site-packages/fugw/solvers/dense.py#521\521]8;;\
           0.27642786502838135                                                                                     

Output()

Output()

[17:04:07] BCD step 14/25  FUGW loss:      0.27215635776519775     Validation loss:                    ]8;id=156185;file:///home/bcl/wanghongyu/.conda/envs/network/lib/python3.11/site-packages/fugw/solvers/dense.py\dense.py]8;;\:]8;id=26398;file:///home/bcl/wanghongyu/.conda/envs/network/lib/python3.11/site-packages/fugw/solvers/dense.py#521\521]8;;\
           0.27215635776519775                                                                                     

Output()

Output()

[17:04:50] BCD step 15/25  FUGW loss:      0.2694956660270691      Validation loss:                    ]8;id=11681;file:///home/bcl/wanghongyu/.conda/envs/network/lib/python3.11/site-packages/fugw/solvers/dense.py\dense.py]8;;\:]8;id=586941;file:///home/bcl/wanghongyu/.conda/envs/network/lib/python3.11/site-packages/fugw/solvers/dense.py#521\521]8;;\
           0.2694956660270691                                                                                      

Output()

Output()

[17:05:34] BCD step 16/25  FUGW loss:      0.26692909002304077     Validation loss:                    ]8;id=85685;file:///home/bcl/wanghongyu/.conda/envs/network/lib/python3.11/site-packages/fugw/solvers/dense.py\dense.py]8;;\:]8;id=731316;file:///home/bcl/wanghongyu/.conda/envs/network/lib/python3.11/site-packages/fugw/solvers/dense.py#521\521]8;;\
           0.26692909002304077                                                                                     

Output()

Output()

[17:06:18] BCD step 17/25  FUGW loss:      0.2654140293598175      Validation loss:                    ]8;id=149721;file:///home/bcl/wanghongyu/.conda/envs/network/lib/python3.11/site-packages/fugw/solvers/dense.py\dense.py]8;;\:]8;id=402122;file:///home/bcl/wanghongyu/.conda/envs/network/lib/python3.11/site-packages/fugw/solvers/dense.py#521\521]8;;\
           0.2654140293598175                                                                                      

Output()

Output()

[17:07:02] BCD step 18/25  FUGW loss:      0.26377835869789124     Validation loss:                    ]8;id=375100;file:///home/bcl/wanghongyu/.conda/envs/network/lib/python3.11/site-packages/fugw/solvers/dense.py\dense.py]8;;\:]8;id=374787;file:///home/bcl/wanghongyu/.conda/envs/network/lib/python3.11/site-packages/fugw/solvers/dense.py#521\521]8;;\
           0.26377835869789124                                                                                     

Output()

Output()

[17:07:46] BCD step 19/25  FUGW loss:      0.26216715574264526     Validation loss:                    ]8;id=426532;file:///home/bcl/wanghongyu/.conda/envs/network/lib/python3.11/site-packages/fugw/solvers/dense.py\dense.py]8;;\:]8;id=210198;file:///home/bcl/wanghongyu/.conda/envs/network/lib/python3.11/site-packages/fugw/solvers/dense.py#521\521]8;;\
           0.26216715574264526                                                                                     

Output()

Output()

[17:08:30] BCD step 20/25  FUGW loss:      0.2614136040210724      Validation loss:                    ]8;id=867942;file:///home/bcl/wanghongyu/.conda/envs/network/lib/python3.11/site-packages/fugw/solvers/dense.py\dense.py]8;;\:]8;id=439898;file:///home/bcl/wanghongyu/.conda/envs/network/lib/python3.11/site-packages/fugw/solvers/dense.py#521\521]8;;\
           0.2614136040210724                                                                                      

Output()

Output()

[17:09:14] BCD step 21/25  FUGW loss:      0.26032334566116333     Validation loss:                    ]8;id=501974;file:///home/bcl/wanghongyu/.conda/envs/network/lib/python3.11/site-packages/fugw/solvers/dense.py\dense.py]8;;\:]8;id=227667;file:///home/bcl/wanghongyu/.conda/envs/network/lib/python3.11/site-packages/fugw/solvers/dense.py#521\521]8;;\
           0.26032334566116333                                                                                     

Output()

Output()

[17:09:58] BCD step 22/25  FUGW loss:      0.25918129086494446     Validation loss:                    ]8;id=145039;file:///home/bcl/wanghongyu/.conda/envs/network/lib/python3.11/site-packages/fugw/solvers/dense.py\dense.py]8;;\:]8;id=431279;file:///home/bcl/wanghongyu/.conda/envs/network/lib/python3.11/site-packages/fugw/solvers/dense.py#521\521]8;;\
           0.25918129086494446                                                                                     

Output()

Output()

[17:10:42] BCD step 23/25  FUGW loss:      0.2586832642555237      Validation loss:                    ]8;id=429401;file:///home/bcl/wanghongyu/.conda/envs/network/lib/python3.11/site-packages/fugw/solvers/dense.py\dense.py]8;;\:]8;id=329276;file:///home/bcl/wanghongyu/.conda/envs/network/lib/python3.11/site-packages/fugw/solvers/dense.py#521\521]8;;\
           0.2586832642555237                                                                                      

Output()

Output()

[17:11:26] BCD step 24/25  FUGW loss:      0.25787296891212463     Validation loss:                    ]8;id=102505;file:///home/bcl/wanghongyu/.conda/envs/network/lib/python3.11/site-packages/fugw/solvers/dense.py\dense.py]8;;\:]8;id=421566;file:///home/bcl/wanghongyu/.conda/envs/network/lib/python3.11/site-packages/fugw/solvers/dense.py#521\521]8;;\
           0.25787296891212463                                                                                     

Output()

Output()

[17:12:10] BCD step 25/25  FUGW loss:      0.2577117383480072      Validation loss:                    ]8;id=739845;file:///home/bcl/wanghongyu/.conda/envs/network/lib/python3.11/site-packages/fugw/solvers/dense.py\dense.py]8;;\:]8;id=650421;file:///home/bcl/wanghongyu/.conda/envs/network/lib/python3.11/site-packages/fugw/solvers/dense.py#521\521]8;;\
           0.2577117383480072                                                                                      

In [55]:
sim_path = Path(f"data/NAPAbench/NAPAbenchVer2/benchmarkDataset/2way/{category}/family1/A-B.sim")

similarity_score = sim_score(sim_path=sim_path, 
coupling=pi,
source_nodes=source_nodes,
target_nodes=target_nodes)

FO_con_soft_score = FO_con_soft(coupling=pi,
source_nodes=source_nodes,
target_nodes=target_nodes,
source_annotations=family.annotations["A"],
target_annotations=family.annotations["B"]
)

FO_con_hard_score1 = FO_con_hard(coupling=pi,
source_nodes=source_nodes,
target_nodes=target_nodes,
source_annotations=family.annotations["A"],
target_annotations=family.annotations["B"],
axis="source"
)

FO_con_hard_score2 = FO_con_hard(coupling=pi,
source_nodes=source_nodes,
target_nodes=target_nodes,
source_annotations=family.annotations["A"],
target_annotations=family.annotations["B"],
axis="target"
)

topk_recall3 = topk_FO_recall(
    pi,
    source_nodes,
    target_nodes,
    family.annotations["A"],
    family.annotations["B"],
    k=3,
)

topk_recall5 = topk_FO_recall(
    pi,
    source_nodes,
    target_nodes,
    family.annotations["A"],
    family.annotations["B"],
    k=5,
)

topk_recall10 = topk_FO_recall(
    pi,
    source_nodes,
    target_nodes,
    family.annotations["A"],
    family.annotations["B"],
    k=10,
)

# Structural metrics
s3_h = S3_hard(family.graphs["A"], family.graphs["B"], pi, source_nodes, target_nodes)
# s3_s = S3_soft(family.graphs["A"], family.graphs["B"], pi, source_nodes, target_nodes)
s3_s=0

print(
    f"SIM={similarity_score:.4f}, FO_cons={FO_con_soft_score:.4f}, {FO_con_hard_score1:.4f}, {FO_con_hard_score2:.4f}, "
    f"Top- FO recall={topk_recall3:.4f}, {topk_recall5:.4f}, {topk_recall10:.4f}, S3_hard={s3_h:.4f}, S3_soft={s3_s:.4f}"
)

SIM=144.6792, FO_cons=0.4574, 0.6170, 0.4758, Top- FO recall=0.6501, 0.6572, 0.6604, S3_hard=0.3127, S3_soft=0.0000


#### CG
- no sim_mat: SIM=193.5201, FO_cons=0.9381, 0.8292, 0.8306, Top- FO recall=0.8477, 0.8486, 0.8486, S3_hard=0.4054, S3_soft=0.0000
- sim_mat: SIM=236.1345, FO_cons=0.9775, 0.9353, 0.9479, Top- FO recall=0.9492, 0.9494, 0.9494, S3_hard=0.4744, S3_soft=0.0000

#### STYCKY
- benchmark: SIM=52.2978, FO_cons=0.3466, 0.3268, 0.2747, Top- FO recall=0.3556, 0.3594, 0.3621, S3_hard=0.3073, S3_soft=0.0000
- rho=1: SIM=37.4984, FO_cons=0.2375, 0.2768, 0.2396, Top- FO recall=0.3036, 0.3097, 0.3148, S3_hard=0.3018, S3_soft=0.0000
- sim_mat: SIM=198.6956, FO_cons=0.6809, 0.7149, 0.5949, Top- FO recall=0.7116, 0.7149, 0.7167, S3_hard=0.3817, S3_soft=0.0000
- rho=1：SIM=144.6792, FO_cons=0.4574, 0.6170, 0.4758, Top- FO recall=0.6501, 0.6572, 0.6604, S3_hard=0.3127, S3_soft=0.0000

#### DMR

#### DMC

#### Tasks
- 测试输入sim_mat但使用均匀分布进行对齐
- 计算最优传输方案的熵